File: Solution/Vishal_Seshagiri_TwitterStreaming.py
import tweepy
from tweepy import Stream
from tweepy.streaming import StreamListener
import re
import json
import random
from collections import Counter

consumer_key = ""
consumer_secret = ""
access_token = ""
access_token_secret = ""

class Listener(StreamListener):
    
	def on_data(self, data):

		global tweet_count
		global window
		global avg_tweet_length
		
		data = json.loads(data)
		
		full_text_retweeted = data.get("retweeted_status")
		
		if None != full_text_retweeted and full_text_retweeted.get("truncated") != False:
			tweet_text = full_text_retweeted.get("extended_tweet").get("full_text")
		else:
			if None != data.get("extended_tweet"):
				tweet_text = data.get("extended_tweet").get("full_text")
			else:
				tweet_text = data.get("text")

		tweet_count += 1
		
		re_hashtag_pattern = "#(\w+)"

		hash_tags_in_current_tweet = re.findall(re_hashtag_pattern, tweet_text)

		if tweet_count <= 100:
			window.append([hash_tags_in_current_tweet, len(tweet_text)])
		else:
			# if 101 in random.sample(range(1, tweet_count), 100):
			if random.randint(1, tweet_count) < 101:
				index_to_replace = random.randint(0, 99)
				window[index_to_replace] = [hash_tags_in_current_tweet, len(tweet_text)]

								
		hashtag_list = []
		tweet_length_sum = 0
		for tweet in window:
			hashtag_list.extend(tweet[0])
			tweet_length_sum += tweet[1]
		
		avg_tweet_length = float(tweet_length_sum) / len(window)

		res = Counter(hashtag_list)
		top5 = "\n".join(["{}:{}".format(i[0], i[1]) for i in res.most_common()[:5]])
		
		if tweet_count > 100:
			print_string = "The number of the twitter from beginning: {tweet_count}\n"\
											"Top 5 hot hashtags:\n{top5}\n"\
											"The average length of the twitter is: {avg_tweet_length}\n"\
											.format(tweet_count=tweet_count, top5=top5, avg_tweet_length=avg_tweet_length)
			
			print(print_string)
		return True
	
	def on_error(self, status):
		print(status)

if __name__=="__main__":

	auth = tweepy.OAuthHandler(consumer_key=consumer_key, consumer_secret=consumer_secret)
	auth.set_access_token(access_token, access_token_secret)
	api = tweepy.API(auth)

	window = []
	tweet_count = 0
	avg_tweet_length = 0.0

	# twitter_stream.disconnect()
	twitter_stream = Stream(auth, Listener(), tweet_mode="extended")
	twitter_stream.filter(track=["#"])

File: Solution/Vishal_Seshagiri_TwitterStreaming.scala
import org.apache.log4j.{Level, Logger}
import org.apache.spark.SparkConf
import org.apache.spark.rdd.RDD
import org.apache.spark.streaming.twitter._
import twitter4j.{FilterQuery, Status}

import scala.util.Random
import org.apache.spark.streaming.{Seconds, StreamingContext}
import org.apache.spark.streaming.dstream.ReceiverInputDStream
import org.scalatest.BeforeAndAfter
import twitter4j.auth.{Authorization, NullAuthorization}

import scala.collection.mutable
import scala.collection.mutable.ListBuffer
import scala.collection.immutable.ListMap

object TwitterStreaming {
  var window = ListBuffer[Tuple2[Array[String], Int]]()
  var tweetCount = 0

  def processor(rdd_statuses : RDD[Status]): Unit = {
    tweetCount += 1
    var statuses = rdd_statuses.map(_.getText).collect().mkString(" ")
    var hashtags = statuses.split(" ").filter(status => status.startsWith("#")).map(h => h.slice(1, h.size).trim).filter(p => !p.isEmpty)
    if (window.size <= 100) {
      window += ((hashtags, statuses.toString().size))
    } else {
//      if (Random.shuffle(1 to tweetCount).take(100).contains(101)){
      if(Random.nextInt((tweetCount)) < 101) {
        var index_to_replace = Random.shuffle(0 to 99).take(1)(0)
        window(index_to_replace) = ((hashtags, statuses.toString().size))
      }
    }

//    var hash_tag_list = ListBuffer[String]()
    var hash_tag_counter = new mutable.HashMap[String, Integer]().withDefaultValue(0)
    var tweet_length_sum = 0
    for(tweet <- window){
      for(hash_tag <- tweet._1){
//        hash_tag_counter.put(hash_tag, counter+1)
        hash_tag_counter(hash_tag) += 1
      }
      tweet_length_sum += tweet._2
    }
    var avg_tweet_length = tweet_length_sum.toFloat / window.size

    var top5 = ListMap(hash_tag_counter.toSeq.sortWith(_._2 > _._2):_*).take(5)
//    top5.map{case (k, v) => k + ":" + v}.mkString("|")

    println(
      s"""The number of twitter from beginning: ${tweetCount}
       |Top 5 hot hashtags: \n${top5.map{case (k, v) => k + ":" + v}.mkString("\n").trim}
       |The average length of the twitter is: ${avg_tweet_length}
       """.stripMargin)
  }

  def main(args : Array[String]): Unit = {

    val consumerKey = ""
    val consumerSecret = ""
    val accessToken = ""
    val accessTokenSecret = ""

    Logger.getLogger("org").setLevel(Level.OFF)
    Logger.getLogger("akka").setLevel(Level.OFF)

    // Authentication of app
    System.setProperty("twitter4j.oauth.consumerKey", consumerKey)
    System.setProperty("twitter4j.oauth.consumerSecret", consumerSecret)
    System.setProperty("twitter4j.oauth.accessToken", accessToken)
    System.setProperty("twitter4j.oauth.accessTokenSecret", accessTokenSecret)

    val sparkConf = new SparkConf().setAppName("TwitterStreaming")
      .setMaster("local[2]")

    val ssc = new StreamingContext(sparkConf, Seconds(2))
    val query = new FilterQuery().language("en").track("#")
    val stream = TwitterUtils.createFilteredStream(ssc, None, Some(query))

//    var statuses = stream.map(_.getText)
    stream.foreachRDD(rdd => processor(rdd))

    ssc.start()
    ssc.awaitTermination()
  }
}
